src/Xamarin.Android.Build.Tasks/Utilities/XAJavaTypeScanner.cs

```diff
@@ -0,0 +1,136 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+
+using Java.Interop.Tools.Cecil;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using Mono.Cecil;
+using Xamarin.Android.Tools;
+
+namespace Xamarin.Android.Tasks;
+
+class JavaType
+{
+	public readonly TypeDefinition Type;
+	public readonly IDictionary<AndroidTargetArch, TypeDefinition>? PerAbiTypes;
+	public bool IsABiSpecific { get; }
+
+	public JavaType (TypeDefinition type, IDictionary<AndroidTargetArch, TypeDefinition>? perAbiTypes)
+	{
+		Type = type;
+		if (perAbiTypes != null) {
+			PerAbiTypes = new ReadOnlyDictionary<AndroidTargetArch, TypeDefinition> (perAbiTypes);
+			IsABiSpecific = perAbiTypes.Count > 1 || (perAbiTypes.Count == 1 && !perAbiTypes.ContainsKey (AndroidTargetArch.None));
+		}
+	}
+}
+
+class XAJavaTypeScanner
+{
+	sealed class TypeData
+	{
+		public readonly TypeDefinition FirstType;
+		public readonly Dictionary<AndroidTargetArch, TypeDefinition> PerAbi;
+
+		public bool IsAbiSpecific => !PerAbi.ContainsKey (AndroidTargetArch.None);
+
+		public TypeData (TypeDefinition firstType)
+		{
+			FirstType = firstType;
+			PerAbi = new Dictionary<AndroidTargetArch, TypeDefinition> ();
+		}
+	}
+
+	public bool ErrorOnCustomJavaObject { get; set; }
+
+	TaskLoggingHelper log;
+	TypeDefinitionCache cache;
+
+	public XAJavaTypeScanner (TaskLoggingHelper log, TypeDefinitionCache cache)
+	{
+		this.log = log;
+		this.cache = cache;
+	}
+
+	public List<JavaType> GetJavaTypes (ICollection<ITaskItem> inputAssemblies, DirectoryAssemblyResolver resolver)
+	{
+		var types = new Dictionary<string, TypeData> (StringComparer.Ordinal);
+		foreach (ITaskItem asmItem in inputAssemblies) {
+			AndroidTargetArch arch = GetTargetArch (asmItem);
+			AssemblyDefinition asmdef = resolver.GetAssembly (asmItem.ItemSpec);
```

Comment:
I suspect/fear that this doesn't do what you think (hope?) it does: 

  * https://github.com/xamarin/java.interop/blob/e1121ead9f1602c1590abfa6235fa16aec567c4d/src/Java.Interop.Tools.Cecil/Java.Interop.Tools.Cecil/DirectoryAssemblyResolver.cs#L207-L210
  * https://github.com/xamarin/java.interop/blob/e1121ead9f1602c1590abfa6235fa16aec567c4d/src/Java.Interop.Tools.Cecil/Java.Interop.Tools.Cecil/DirectoryAssemblyResolver.cs#L252-L279

In particular, note lines 254 and 257, which uses the "assembly name" -- which is based off the the filename *without extension* -- as the key into `DirectoryAssemblyResolver.cache`.

TL;DR: this will return the *first* assembly encountered, regardless of ABI, always.  `arch` (previous line)` is not relevant.

`DirectoryAssemblyResolver` will not help you here, unless (maybe?) you have *per-ABI* `DirectoryAssemblyResolver` instances?

Reply:
Previous code took 30s to load an assembly, so I switched to this. It seems the next step is to add a new resolver. Next week, then.